lib/srv/discovery/kube_integration_watcher.go

```diff
@@ -0,0 +1,295 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discovery
+
+import (
+	"context"
+	"errors"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+	"github.com/gravitational/teleport/lib/automaticupgrades/version"
+	"github.com/gravitational/teleport/lib/integrations/awsoidc"
+	"github.com/gravitational/teleport/lib/srv/discovery/common"
+)
+
+func (s *Server) startKubeIntegrationWatchers() error {
+	if s.dynamicMatcherWatcher == nil {
+		return nil
+	}
+
+	var mu sync.Mutex
+	// enrollingClusters keeps track of clusters that are in the process of being enrolled, so they are
+	// not yet among existing clusters, but we also should not try to enroll them again.
+	enrollingClusters := map[string]bool{}
+
+	clt := s.AccessPoint
+
+	pong, err := clt.Ping(s.ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	proxyPublicAddr := pong.GetProxyPublicAddr()
+
+	releaseChannels := automaticupgrades.Channels{}
+	if err := releaseChannels.CheckAndSetDefaults(s.ClusterFeatures); err != nil {
+		return trace.Wrap(err)
+	}
+
+	watcher, err := common.NewWatcher(s.ctx, common.WatcherConfig{
+		FetchersFn:     s.getKubeIntegrationFetchers,
+		Log:            s.Log.WithField("kind", types.KindKubernetesCluster),
+		DiscoveryGroup: s.DiscoveryGroup,
+		Interval:       s.PollInterval,
+		Origin:         types.OriginCloud,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	go watcher.Start()
+
+	go func() {
+		for {
+			select {
+			case resources := <-watcher.ResourcesC():
+				if len(resources) == 0 {
+					continue
+				}
+
+				existingServers, err := clt.GetKubernetesServers(s.ctx)
```

Comment:
I think it's easier to separate both old and new method so you don't need to touch GetKubernetesClusters when using the new fetcher. the new fetcher should install the agent but shouldn't create the kube cluster otherwise someone else might be watching those resources and failing

Reply:
I get kubernetes clusters (not creating though) just so we don't try to install agent on a cluster that is already present, maybe it was added manually or something 🤷